### PR TITLE
bad location protection

### PR DESCRIPTION
**Pull request by whit537**

### CODE REVIEW

**Comment by whit537** [comment]
Ready for review.

File: aspen/website.py [hunk 1]
```diff
@@ -68,15 +69,19 @@ def redirect(self, location, code=None, permanent=False, base_url=None, response
         permanent is True and 302 (Found) if it is False. If url doesn't start
         with base_url (defaulting to self.base_url), then we prefix it with
         base_url before redirecting. This is a protection against open
-        redirects. If you provide your own response we will set .code and
-        .headers['Location'] on it.
+        redirects. If base_url is not '' and location is not an absolute path,
```
Comment (pjz) [inline aspen/website.py]: How about "If location is a relative path but base_url is empty, we raise BadLocation."

Comment (whit537) [inline aspen/website.py]: I think we want to allow relative path redirects, in which case, with a non-empty `website.base_url`, the `base_url` parameter becomes the way to explicitly whitelist a relative path or absolute URL redirect. See https://github.com/gratipay/gratipay.com/commit/ebce89f10244a65d52e8eb0c07a015d8eea7019a for relative path example usage, and https://github.com/gratipay/gratipay.com/commit/fca0cf1e4d990db6df7c80901541e6022d53013e for absolute URL.

Comment (pjz) [inline aspen/website.py]: correct; I just meant to rewrite the awkward sentence. How about: "base_url cannot be empty if you wish to use relative paths; if you try, BadLocation will be raised."

Comment (whit537) [inline aspen/website.py]: Well, but I think you've still got the logic backwards: `base_url` cannot *not* be empty if you wish to use relative paths. If you pass a relative path such as `../foo` for location and a non-empty `base_url` such as `http://example.com`, then you're asking for a computed location of `http://example.com../foo`, which is garbage. No?

Instead, you should override any globally configured `base_url` by passing `base_url=''` in a specific `redirect` call.

```python
website.base_url = 'http://example.com'
website.redirect('../foo', base_url='')
```

Comment (pjz) [inline aspen/website.py]: Ah! So then how about "base_url must be empty if you wish to use a relative path as location; if it's not, a BadLocation error will be thrown." ?  It's much clearer, IMO, than "If base_url is not '' and location is not an absolute path, we raise BadLocation"

Comment (whit537) [inline aspen/website.py]: Fair enough. How's b2ca63e030c02e75d026fa6d73f7eb260aa23523 look?

**Pull request merged**